microsoft/PowerToys
# KBM - Set up tests for keyboard hook remapping logic

## Summary of the Pull Request
Added tests for keyboard manager to cover its hook remapping logic for Single key remaps and OS level shortcut remaps. KBM now has 28 tests
![image](https://user-images.githubusercontent.com/32061677/83686492-248fac00-a59f-11ea-8cca-64c7527a3600.png)

In order to test hook based logic, a mock version of keyboard input had to be created which simulates low level hook behavior. This can be moved to common lib in a future PR in case it could be useful to other PowerToys.


## PR Checklist
* [X] Applies to #3112 


## Detailed Description of the Pull Request / Additional comments
- For testing the keyboard remapping logic without actually modifying keyboard states it was necessary to mock functions like SendInput, GetAsyncKeyState and also low level hook functionality. For C++ mocking, based on [this article](https://games.greggman.com/game/using-a-mock-library-to-make-unit-testing-easier-in-c/), some of the code was refactored such that SendInput and GetAsyncKeyState functions were wrapped into a class derived from an interface. On the test project side this interface was implemented with mocked functions.
- The mocked input was simulated by storing a keyboard state variables (stores true or false for all 256 virtual key codes), and SendInput was mocked such that it iterates over the key events, and calls the hook on each of them, and it doesn't  set the keyboard state variable if the hook suppresses the event. This logic was done to simulate how SendInput and low level hooks actually work.
- In order to keep the MockedInput, Input and InputInterface classes as reusable as possible, under MockedInput rather than hardcoding in the low level hook procedure, it was set such that there is a reference to an std::function which can be set at runtime. Thus it is possible to test code without any hook as well.
- For making tests independent, at the start of each test the KeyboardState variable is reset, hook is set to null and all remappings are cleared.

## Tests Added
- **Mocked Input Sanity Test** - To ensure that the mocked input infrastructure is working, we check the keyboard state before and after a key event.
- **Single Key Remapping:**
    - Basic test to check if target key is set for a remapped key
    - If a key is remapped to disabled, then it shouldn't change the keyboard state (this exists in the backend but hasn't been exposed to the UI as a feature yet)
    - If a key is remapped to Win (Both) it should set the Win (left) key state (since Win Both is not an actual key code, its created by us)
- **OS Level Shortcut Remapping:**
    - The following cases were tested for all the scenarios - i.e. on shortcut key down, on pressing and releasing the shortcut, and separate cases were made for shortcuts with the same modifiers, different modifiers, or some common modifiers
        - Two key shortcut to two key shortcut
        - Two key shortcut to three key shortcut
        - Three key shortcut to two key shortcut
        - Three key shortcut to three key shortcut
    - If a two key shortcut is remapped, invoking a three key shortcut containing that should not get remapped. Ex: Ctrl+A->Alt+D, Ctrl+Shift+A should not be affected
    - If a three key shortcut is remapped, invoking a two key shortcut which is a subset of that should not get remapped. Ex: Ctrl+Shift+A->Alt+D, Ctrl+A should not be affected
    - If a shortcut is remapped, and after invoking it you press another action key without releasing the modifiers, it should reset back to the physical keys being pressed (if they are not remapped). Ex: Ctrl+A remapped to Alt+D. User presses Ctrl+A, releases A and then presses C, keyboard should detect it as Ctrl+C and do Copy action.
    - If a shortcut with Win (Both) is remapped, it should work with both Win (Left) or Win (Right). (Test added since Win (Both) isn't an actual key code)
    - If a shortcut with Win (Both) is remapped, and after invoking it with Win (Left/Right) you press another action key without releasing the modifiers, it should reset back to the physical keys being pressed, i.e. the same win key which was pressed (L/R).
    - Test for Alt+Tab fix behavior (#3965) - If a shortcut is remapped, and you invoke it but only let go of the action key, the modifier should remain pressed until a key not present in the shortcut is pressed. Required to test that remaps to Alt+Tab behave correctly.
## Pending Work
- Integration tests for interaction between single key remaps and shortcut remaps. To be tracked in a separate issue.


## Validation Steps Performed
Build and run KBM in debug/release to ensure all functionality works. Tests pass in both debug and release.
